spinnman/spalloc/spalloc.py

```diff
@@ -0,0 +1,822 @@
+# Copyright (c) 2021-2022 The University of Manchester
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+Implementation of the client for the Spalloc web service.
+"""
+
+from logging import getLogger
+from multiprocessing import Process, Queue
+from packaging.version import Version
+import queue
+import requests
+import sqlite3
+import struct
+import threading
+from typing import Dict, List, Tuple
+from websocket import WebSocket
+from spinn_utilities.abstract_base import AbstractBase, abstractmethod
+from spinn_utilities.abstract_context_manager import AbstractContextManager
+from spinn_utilities.log import FormatAdapter
+from spinn_utilities.overrides import overrides
+from spinnman.connections.abstract_classes import (
+    Connection, Listenable)
+from spinnman.constants import SCP_SCAMP_PORT, UDP_BOOT_CONNECTION_DEFAULT_PORT
+from spinnman.exceptions import SpinnmanTimeoutException
+from .enums import SpallocState, ProxyProtocol
+from .session import Session, SessionAware
+from .utils import parse_service_url, get_hostname
+from .abstract_classes import (
+    AbstractSpallocClient, SpallocProxiedConnection,
+    SpallocSCPConnection, SpallocBootConnection, SpallocJob,
+    SpallocMachine, SpallocEIEIOConnection, SpallocEIEIOListener)
+from spinnman.transceiver import Transceiver
+
+logger = FormatAdapter(getLogger(__name__))
+_open_req = struct.Struct("<IIIII")
+_close_req = struct.Struct("<III")
+_open_listen_req = struct.Struct("<II")
+# Open and close share the response structure
+_open_close_res = struct.Struct("<III")
+_open_listen_res = struct.Struct("<IIB4I")
+_msg = struct.Struct("<II")
+_msg_to = struct.Struct("<IIII")
+
+
+class SpallocClient(AbstractContextManager, AbstractSpallocClient):
```

Comment:
Maybe rename this spalloc_client.py as well.  Although it has other classes in the file, those are private to this module and so don't need to be broken out, but the main thing in this file is SpallocClient.

Reply:
Done